tt_metal/impl/buffers/buffer.cpp

```diff
@@ -267,11 +285,52 @@ Buffer &Buffer::operator=(Buffer &&other) {
 }
 
 void Buffer::allocate() {
+    if(allocation_status_ != AllocationStatus::NOT_ALLOCATED) {
+        return;
+    }
+
+    allocation_status_ = AllocationStatus::ALLOCATION_REQUESTED;
+
     TT_ASSERT(this->device_ != nullptr);
     // L1 and Trace buffers (which live in DRAM) are allocated top down!
     bool bottom_up = this->bottom_up_.value_or(this->is_dram());
     detail::AllocateBuffer(this, bottom_up);
     detail::BUFFER_MAP.insert({this->device_->id(), this->address_}, this);
+
+    allocation_status_ = AllocationStatus::ALLOCATED;
+}
+
+void Buffer::deallocate() {
+    if (allocation_status_ == AllocationStatus::NOT_ALLOCATED) {
+        return;
+    }
+
+    if (this->device_ == nullptr
+        or not this->device_->initialized_
+        or this->size_ == 0) {
+        return;
+    }
+
+    TT_ASSERT(this->device_->allocator_ != nullptr, "Expected allocator to be initialized!");
+    detail::BUFFER_MAP.erase({this->device_->id(), this->address_});
+    detail::DeallocateBuffer(this);
+
+    allocation_status_ = AllocationStatus::NOT_ALLOCATED;
+}
+
+Buffer::AllocationStatus Buffer::get_allocation_status()
+{
+    return allocation_status_;
+}
+
+uint32_t Buffer::address() const {
```

Comment:
Important part of the contract update here - address can't be access from non-worker thread.
Good implementation could just block till allocation is completed. 
For this we could merge ttnn buffer and metal buffer moving push_work to metal buffer's allocate/dellocate.

Reply:
The best solution probably completely avoids buffer's self allocation, so that allocation always happens via allocator. in that case allocator can return a ready buffer if sync allocate is called or something like a buffer promise if async allocate is called. Fantasizing out loud.